codegen/module.go

```diff
@@ -574,92 +566,182 @@ func (system *ModuleSystem) ResolveModules(
-				if err != nil {
-					return nil, errors.Wrapf(
-						err,
-						"internal error: cannot make %q relative to %q",
-						moduleDirAbs,
-						baseDirectory,
-					)
-				}
+type resolveResult struct {
+	err    error
+	module map[string]map[string]*ModuleInstance
+}
 
-				classConfigPath, _, _ := getConfigFilePath(moduleDirAbs, className)
-				if classConfigPath == "" {
-					fmt.Printf("  no class config found in %s directory\n", moduleDirAbs)
-					// No class config found in this directory, skip over it
-					continue
-				}
+func (system *ModuleSystem) resolveModule(baseDirectory string, moduleDirectoryGlob string, className string,
+	packageRoot string, targetGenDir string, defaultDependencies []ModuleDependency,
+	ch chan resolveResult, wg *sync.WaitGroup, parentModule map[string]map[string]*ModuleInstance) {
+	defer wg.Done()
+	curModules := map[string]map[string]*ModuleInstance{}
+	mergeResolveMap(parentModule, curModules, "")
 
-				instance, instanceErr := system.readInstance(
-					className,
-					packageRoot,
-					baseDirectory,
-					targetGenDir,
-					moduleDir,
+	moduleDirectoriesAbs, err := filepath.Glob(filepath.Join(baseDirectory, moduleDirectoryGlob))
+	if err != nil {
+		ch <- resolveResult{err: errors.Wrapf(err, "error globbing %q", moduleDirectoryGlob)}
+		return
+	}
+
+	for _, moduleDirAbs := range moduleDirectoriesAbs {
+		stat, err := os.Stat(moduleDirAbs)
+		if err != nil {
+			ch <- resolveResult{
+				err: errors.Wrapf(
+					err,
+					"internal error: cannot stat %q",
 					moduleDirAbs,
-					defaultDependencies,
-				)
-				if instanceErr != nil {
-					return nil, errors.Wrapf(
-						instanceErr,
-						"Error reading multi instance %q",
-						moduleDir,
-					)
-				}
+				),
+			}
+			return
+		}
 
-				resolvedModules[instance.ClassName] = append(resolvedModules[instance.ClassName], instance)
+		if !stat.IsDir() {
+			// If a *-config.yaml file, or any other metadata file also matched the glob, skip it, since we are
+			// interested only in the containing directories.
+			continue
+		}
 
-				if className != instance.ClassName {
-					return nil, fmt.Errorf(
-						"invariant: all instances in a multi-module directory %q are of type %q (violated by %q)",
-						moduleDirectoryGlob,
-						className,
-						moduleDir,
-					)
-				}
-			}
+		moduleDir, err := filepath.Rel(baseDirectory, moduleDirAbs)
+		if err != nil {
+			ch <- resolveResult{err: errors.Wrapf(
+				err,
+				"internal error: cannot make %q relative to %q",
+				moduleDirAbs,
+				baseDirectory,
+			)}
+			return
+		}
 
-			// Resolve dependencies for all classes
-			resolveErr := system.populateResolvedDependencies(
-				resolvedModules[className],
-				resolvedModules,
-			)
-			if resolveErr != nil {
-				return nil, resolveErr
-			}
+		classConfigPath, _, _ := getConfigFilePath(moduleDirAbs, className)
+		if classConfigPath == "" {
+			fmt.Printf("  no class config found in %s directory\n", moduleDirAbs)
```

Comment:
Did you intend to leave this print line here? or was it just for debug

Reply:
code-gen prints out the lines to console for debugging purposes